components/tidb_query_datatype/src/codec/mysql/json/json_contains.rs

```diff
@@ -0,0 +1,106 @@
+// Copyright 2022 TiKV Project Authors. Licensed under Apache-2.0.
+
+use std::cmp::Ordering;
+
+use super::{super::Result, JsonRef, JsonType};
+
+impl<'a> JsonRef<'a> {
+    /// `json_contains` is the implementation for JSON_CONTAINS in mysql
+    /// <https://dev.mysql.com/doc/refman/5.7/en/json-search-functions.html#function_json-contains>
+    /// See `ContainsBinaryJSON()` in TiDB `types/json_binary_functions.go`
+    pub fn json_contains(&self, target: JsonRef<'_>) -> Result<bool> {
+        match self.type_code {
+            JsonType::Object => {
+                if target.type_code == JsonType::Object {
+                    let elem_count = target.get_elem_count();
+                    for i in 0..elem_count {
+                        let key = target.object_get_key(i);
+                        let val = target.object_get_val(i)?;
+                        let idx = self.object_search_key(key);
+                        match idx {
+                            None => {
+                                return Ok(false);
+                            }
+                            Some(idx) => {
+                                let exp = self.object_get_val(idx)?;
+                                if !(exp.json_contains(val)?) {
+                                    return Ok(false);
+                                }
+                            }
+                        }
+                    }
+                    return Ok(true);
+                }
+            }
+            JsonType::Array => {
+                if target.type_code == JsonType::Array {
+                    let elem_count = target.get_elem_count();
+                    for i in 0..elem_count {
+                        if !(self.json_contains(target.array_get_elem(i)?)?) {
+                            return Ok(false);
+                        }
+                    }
+                    return Ok(true);
+                }
+                let elem_count = self.get_elem_count();
+                for i in 0..elem_count {
+                    if self.array_get_elem(i)?.json_contains(target)? {
+                        return Ok(true);
+                    }
+                }
+            }
+            _ => {
+                return match self.partial_cmp(&target).unwrap() {
```

Comment:
Was it possible that self and target are not comparable for example  self is a string while target is an array?

Reply:
Or was it possible that `self.partial_cmp(&target) == None`

Reply:
> Was it possible that self and target are not comparable for example self is a string while target is an array?

The json comparison follows a specific precedence. For example, arrays are always greater than string.

> Or was it possible that `self.partial_cmp(&target) == None`

It's `None` when the data is corrupted. For example, if a json object declares itself as a string, but the value buffer is shorter than 8 (the size of u64).

I don't know how TiKV should behave when the data is corrupted :thinking: . What's your opinion?

Reply:
But actually, if the data is corrupted, `self.partial_cmp` this function itself can panic. Because if the string declares its length as 4, but the buffer is shorter than 8+4=12, it will panic because of indexing out of range.

Reply:
@AndreMouche I think it's fine to merge this PR. It seems that both TiDB and TiKV cannot handle this situation and will panic directly. I've created an issue to track this problem: https://github.com/tikv/tikv/issues/13592. If you have any opinion related with this scenario, feel free to discuss under that issue.